src/expr/macro/src/context.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use proc_macro2::TokenStream;
+use quote::{quote, quote_spanned, ToTokens};
+use syn::parse::{Parse, ParseStream};
+use syn::{Error, FnArg, Ident, ItemFn, Result, Token, Type, Visibility};
+
+/// See [`super::define_context!`].
+#[derive(Debug, Clone)]
+pub(super) struct DefineContextField {
+    vis: Visibility,
+    name: Ident,
+    ty: Type,
+}
+
+/// See [`super::define_context!`].
+#[derive(Debug, Clone)]
+pub(super) struct DefineContextAttr {
+    fields: Vec<DefineContextField>,
+}
+
+impl Parse for DefineContextField {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let vis: Visibility = input.parse()?;
+        let name: Ident = input.parse()?;
+        input.parse::<Token![:]>()?;
+        let ty: Type = input.parse()?;
+
+        Ok(Self { vis, name, ty })
+    }
+}
+
+impl Parse for DefineContextAttr {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let fields = input.parse_terminated(DefineContextField::parse, Token![,])?;
+        Ok(Self {
+            fields: fields.into_iter().collect(),
+        })
+    }
+}
+
+impl DefineContextField {
+    pub(super) fn gen(self) -> Result<TokenStream> {
+        let Self { vis, name, ty } = self;
+
+        {
+            let name_s = name.to_string();
+            if name_s.to_uppercase() != name_s {
+                return Err(Error::new_spanned(
+                    name,
+                    "the name of context variable should be uppercase",
+                ));
+            }
+        }
+
+        Ok(quote! {
+            #[allow(non_snake_case)]
+            pub mod #name {
+                use super::*;
+                pub type Type = #ty;
+
+                tokio::task_local! {
+                    static LOCAL_KEY: #ty;
+                }
+
+                #vis fn try_with<F, R>(f: F) -> Result<R, risingwave_expr::ExprError>
+                where
+                    F: FnOnce(&#ty) -> R
+                {
+                    LOCAL_KEY.try_with(f).map_err(|_| risingwave_expr::ContextUnavailable::new(stringify!(#name))).map_err(Into::into)
```

Comment:
We may need the function name in error message as well:
```
dev=> set query_mode to distributed;
SET_VARIABLE
dev=> select ''::regclass;
ERROR:  QueryError: Expr error: Context DB_NAME not found
```
Or maybe it can be part of any `ExprError`?

Reply:
It should be a part of any ExprError